x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/action/enrollment/TransportCreateEnrollmentTokenAction.java

```diff
@@ -0,0 +1,147 @@
+                             ActionListener<CreateEnrollmentTokenResponse> listener) {
+        try {
+            final KeyConfig keyConfig = sslService.getHttpTransportSSLConfiguration().keyConfig();
+            if (keyConfig instanceof StoreKeyConfig == false) {
+                listener.onFailure(new IllegalStateException(
+                    "Unable to create an enrollment token. Elasticsearch node HTTP layer SSL configuration is not configured with a " +
+                        "keystore"));
+                return;
+            }
+            final List<Tuple<PrivateKey, X509Certificate>> httpCaKeysAndCertificates =
+                ((StoreKeyConfig) keyConfig).getPrivateKeyEntries(environment).stream()
+                    .filter(t -> t.v2().getBasicConstraints() != -1).collect(Collectors.toList());
+            if (httpCaKeysAndCertificates.isEmpty()) {
+                listener.onFailure(new IllegalStateException(
+                    "Unable to create an enrollment token. Elasticsearch node HTTP layer SSL configuration Keystore doesn't contain any " +
+                        "PrivateKey entries where the associated certificate is a CA certificate"));
+                return;
+            } else if (httpCaKeysAndCertificates.size() > 1) {
+                listener.onFailure(new IllegalStateException(
+                    "Unable to create an enrollment token. Elasticsearch node HTTP layer SSL configuration Keystore contain multiple " +
+                        "PrivateKey entries where the associated certificate is a CA certificate"));
+                return;
+            } else {
+                final TimeValue expiration = TimeValue.timeValueSeconds(ENROLL_API_KEY_EXPIRATION_SEC);
+                final String[] clusterPrivileges = { "enroll" };
+                final List<RoleDescriptor> roleDescriptors = List.of(new RoleDescriptor("create_enrollment_token", clusterPrivileges,
+                    null, null));
+                CreateApiKeyRequest apiRequest = new CreateApiKeyRequest("enrollment_token_API_key_" + UUIDs.base64UUID(),
+                    roleDescriptors, expiration);
+                final String fingerprint = SslUtil.calculateFingerprint(httpCaKeysAndCertificates.get(0).v2());
+                generator.generateApiKey(securityContext.getAuthentication(), apiRequest,
+                    ActionListener.wrap(
+                        CreateApiKeyResponse -> {
+                            try {
+                                final String address = nodeService.info(false, false, false, false, false, false,
+                                    true, false, false, false, false).getInfo(HttpInfo.class).getAddress().publishAddress().toString();
+                                final XContentBuilder builder = JsonXContent.contentBuilder();
+                                builder.startObject();
+                                builder.field("adr", address);
```

Comment:
I think there is no guarantee that enrolment mode is enabled on other nodes?

Reply:
Yes, that's true, it's rather complex to deal with this.

In addition, users of "Security ON by default" (most) most likely won't have complex HTTP publish addresses (though it is possible), so that some nodes are reachable and others are not (these complex setups probably have specific certificates infrastructure).

On further thought, my proposal sounds like over-engineering, I'm going to retract it, thanks Yang!